### PR TITLE
httpbakery/agent: support serialization

### PR DESCRIPTION
**Pull request by rogpeppe**
Also make key pair unmarshaling a little more robust in bakery.

### CODE REVIEW

File: bakery/keys.go [hunk 1]
```diff
@@ -136,6 +142,36 @@ type KeyPair struct {
 	Private PrivateKey `json:"private"`
 }
 
+// UnmarshalJSON implements json.Unmarshaler.
+func (k *KeyPair) UnmarshalJSON(data []byte) error {
+	type keyPair KeyPair
+	if err := json.Unmarshal(data, (*keyPair)(k)); err != nil {
+		return err
+	}
+	if k.Public.isZero() {
```
Comment (frankban) [inline bakery/keys.go]: `return checkZero(k)` also reused by UnmarshalYAML below?

File: bakery/keys_test.go [hunk 2]
```diff
@@ -46,20 +53,74 @@ func (*KeysSuite) TestKeyPairMarshalJSON(c *gc.C) {
...
+	err := json.Unmarshal([]byte(data), &k)
+	c.Assert(err, gc.ErrorMatches, `missing public key`)
+}
+
+func (*KeysSuite) TestKeyPairUnmarshalJSONMissingPrivateKey(c *gc.C) {
+	data := `{"public": "7ZcOvDAW9opAIPzJ7FdSbz2i2qL8bFZapDlmNLpMzpU="}`
+	var k bakery.KeyPair
+	err := json.Unmarshal([]byte(data), &k)
+	c.Assert(err, gc.ErrorMatches, `missing private key`)
+}
```
Comment (frankban) [inline bakery/keys_test.go:104]: Worth adding a test with 
```
data := `{"private": "", "public: ""}
```
?

File: httpbakery/agent/cookie.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+// ErrNoAgentLoginCookie is the error returned when the expected
+// agent login cookie has not been found.
+var ErrNoAgentLoginCookie = errgo.New("no agent-login cookie found")
+
+// LoginCookie returns details of the agent login cookie
+// from the given request. If no agent-login cookie is found,
+// it returns an ErrNoAgentLoginCookie error.
+func LoginCookie(req *http.Request) (username string, key *bakery.PublicKey, err error) {
+	c, err := req.Cookie(cookieName)
+	if err != nil {
```
Comment (frankban) [inline httpbakery/agent/cookie.go:50]: Shoudl we check for ErrNoCookie? I know it's the only possible error, so [shrug].

**Review by frankban — APPROVED** [review]
Looks good with only minor observations, thank you!

File: httpbakery/agent/cookie.go (hunk 3 above)
Comment (rogpeppe) [reply httpbakery/agent/cookie.go:50]: Possibly. This code isn't new (it just moved) so if we cared, it should probably be done in another PR.

File: bakery/keys.go (hunk 1 above)
Comment (rogpeppe) [reply bakery/keys.go]: done

File: bakery/keys_test.go (hunk 2 above)
Comment (rogpeppe) [reply bakery/keys_test.go:104]: done (and another test)

File: bakery/keys_test.go [hunk 4]
```diff
@@ -38,6 +39,12 @@ func (*KeysSuite) TestMarshalText(c *gc.C) {
 	c.Assert(key1, gc.Equals, testKey)
 }
 
+func (*KeysSuite) TestUnmarshalTextWrongKeyLength(c *gc.C) {
+	var key bakery.Key
+	err := key.UnmarshalText([]byte("aGVsbG8K"))
+	c.Assert(err, gc.ErrorMatches, `wrong length for base64 key, got 6 want 32`)
```
Comment (mhilton) [inline bakery/keys_test.go]: The word base64 in this error confuses me, especially as it is 8 long, not 6.

File: httpbakery/agent/serialize.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+var _ interface {
+	json.Unmarshaler
+	json.Marshaler
+	yaml.Unmarshaler
+	yaml.Marshaler
+} = (*Visitor)(nil)
+
+// UnmarshalJSON implements JSON.Unmarshaler.
```
Comment (mhilton) [inline httpbakery/agent/serialize.go]: json.Unmarshaler

**Review by mhilton — APPROVED** [review]
LGTM with a couple of thoughts.

File: bakery/keys_test.go (hunk 4 above)
Comment (rogpeppe) [reply bakery/keys_test.go]: changed

File: httpbakery/agent/serialize.go (hunk 5 above)
Comment (rogpeppe) [reply httpbakery/agent/serialize.go]: done

**Pull request merged**